### ISSUE TITLE
Uploading stemcell fails with Error 100: Unknown CPI error 'Unknown' with message 'connect_write timeout reached'

### ISSUE THREAD

**Issue by MoizArif** [cloudfoundry-incubator/bosh-openstack-cpi-release#44]
I am trying to upload the bosh stemcell image and it keeps on failing with this message:

`Error 100: Unknown CPI error 'Unknown' with message 'connect_write timeout reached'`

The complete output of the bosh upload stemcell command can be seen here: http://pastebin.com/30bnXAtc

The debug output of this task can be seen here: http://pastebin.com/ZUi4RMQY

i have tried many stemcell images but all get stuck at 96% and give the same error as above. Anyone knows how to solve this?

**Comment by Kiemes** [comment]
Hi @MoizArif,

thanks for your question. Apparently your Glance was not able to handle the connection in time. The default connection timeout is 60s.
We have the possibility of setting `connect_timeout` in the [connection_options](http://bosh.io/jobs/openstack_cpi?source=github.com/cloudfoundry-incubator/bosh-openstack-cpi-release&version=25#p=openstack.connection_options) of your bosh director deployment manifest. Generally, all [Excon connection options](https://github.com/excon/excon#options) are available. We will improve the documentation here, explicitly mentioning that fact.

Your debug output shows that you are using the default `connect_timeout` with 60 seconds (:connect_timeout=>60).

Please try again with a higher timeout and let us know if it worked.
Could you also share with us on which OpenStack version you are running and which provider you are using?

Best regards,
 Tom

**Comment by MoizArif** [comment]
Hi @Kiemes 
Thanks for the reply. 

I am using OpenStack Kilo deployed through MOS. 

I have added the tag "write_timeout: 360" under the openstack section in bosh.yml file and redeployed the director but still facing the same issue. 

May be i am adding the timeout in an incorrect fashion. Uploading the stemcell yields the same error as before.

**Comment by Kiemes** [comment]
@MoizArif `connect_timeout` would be the right key in your case. Make sure your openstack looks similar to this:

```yaml
openstack: &openstack
      auth_url: IDENTITY-API-ENDPOINT # <--- Replace with OpenStack Identity API endpoint
      tenant: OPENSTACK-TENANT # <--- Replace with OpenStack tenant name
      username: OPENSTACK-USERNAME # <--- Replace with OpenStack username
      api_key: OPENSTACK-PASSWORD # <--- Replace with OpenStack password
      default_key_name: bosh
      default_security_groups: [bosh]
      connection_options:
         connect_timeout: 360    
```

**Comment by MoizArif** [comment]
@Kiemes I made the necessary changes but the uploading process is still getting stuck at 96%. But this time the timeout seemed to have worked and therefore i got another error now:

```
Error 100: Unable to connect to the OpenStack Keystone API http://10.10.26.2:5000/v2.0/tokens
Connection timed out - connect(2) for 10.10.26.2:5000 (Errno::ETIMEDOUT)
```

From the logs:
```
E, [2016-07-11T15:34:15.303078 #24198] ERROR -- : Failed 2 times, last failure due to: #<Excon::Errors::SocketError: Connection timed out - connect(2) for 10.10.26.2:5000 (Errno::ETIMEDOUT)>
[fog][WARNING] Unrecognized arguments: provider
```

However, my OpenStack auth_url is working when i execute openstack cli commands. 
`export OS_AUTH_URL='http://10.10.26.2:5000/v2.0/'`

**Comment by loewenstein** [comment]
@MoizArif Could you please send us your bosh.yml and the recent version of the debug log for the upload stemcell task.
Don't forget to redact any secrets.

In parallel let's see what the openstack cli equivalent does.
Could you please try to upload the stemcell 'manually' like 
1. Untar stemcell
2. Open stemcell.MF in text editor (for looking up '<manifest value>' below)
3. `openstack image create --file <path-to-untared stemcell>/root.img --disk-format <manifest value> --property version=<manifest value> --property hypervisor_type=<manifest value: 'hypervisor'> --container-format <manifest value> --property os_type=<manifest value> --property os_distro=<manifest value> --property architecture=<manifest value> --property auto_disk_config=<manifest value> <name>`

Is this successful?

**Comment by MoizArif** [comment]
@loewenstein 

1. Here is my bosh.yml file: http://pastebin.com/qDPMZv9A
2. Yes, the manual upload of the image is successful, the image is active. Used the following command:

`glance image-create --file image --disk-format qcow2 --property version='3262.2' --property hypervisor_type=kvm --container-format bare --property os_type=linux --property os_distro=ubuntu --property architecture=x86_64 --property auto_disk_config=true --name bosh-openstack-kvm-ubuntu-trusty-go_agent`

**Comment by voelzmo** [comment]
@MoizArif: It seems the Director VM is not able to contact your OpenStack endpoints. Maybe due to security group settings or network restrictions.

Please try the above commands from the Director VM to eliminate those issues.

**Comment by MoizArif** [comment]
@voelzmo : It seems that OpenStack clients are installed in the BOSH director VM. Is there any other way of executing the above command from inside the Director VM?


The security group used is bosh and allows all traffic in & out of the VM. I have attached the screenshot of the security group rules i have configured. 

![image](https://cloud.githubusercontent.com/assets/3896078/16774714/aec555ee-4876-11e6-8bd1-c6bf92572ee9.png)

**Comment by MoizArif** [comment]
Hi guys, 
Since keystone internal URL wasn't accessible from the Director VM, i used the keystone public url and used it as auth_url under openstack tag in bosh.yml and then redeployed the director VM. Note that the public url is on the same network as the floating IP assigned to the director VM and the public url is also pingable. Now when i upload the stemcell it can access the url but now it gives me a different error related to certificates:

```
`bosh-stemcell:  96% |ooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo     | 510.6MB  27.5MB/s ETA:  00:00:00
Director task 3
  Started update stemcell
  Started update stemcell > Extracting stemcell archive. Done (00:00:06)
  Started update stemcell > Verifying stemcell manifest. Done (00:00:00)
  Started update stemcell > Checking if this stemcell already exists. Done (00:00:00)
  Started update stemcell > Uploading stemcell bosh-openstack-kvm-ubuntu-trusty-go_agent/3262.2 to the cloud. Failed: Unable to connect to the OpenStack Keystone API https://public.fuel.local:5000/v2.0/tokens
SSL_connect returned=1 errno=0 state=SSLv3 read server certificate B: certificate verify failed (OpenSSL::SSL::SSLError) Unable to verify certificate. This may be an issue with the remote host or with Excon.Excon has certificates bundled, but these can be customized.`Excon.defaults[:ssl_ca_path] = path_to_certs`, `ENV['SSL_CERT_DIR'] = path_to_certs`, `Excon.defaults[:ssl_ca_file] = path_to_file`, `ENV['SSL_CERT_FILE'] = path_to_file`, `Excon.defaults[:ssl_verify_callback] = callback` (see OpenSSL::SSL::SSLContext#verify_callback), or `Excon.defaults[:ssl_verify_peer] = false` (less secure). (00:00:12)
Error 100: Unable to connect to the OpenStack Keystone API https://public.fuel.local:5000/v2.0/tokens
SSL_connect returned=1 errno=0 state=SSLv3 read server certificate B: certificate verify failed (OpenSSL::SSL::SSLError) Unable to verify certificate. This may be an issue with the remote host or with Excon.Excon has certificates bundled, but these can be customized.`Excon.defaults[:ssl_ca_path] = path_to_certs`, `ENV['SSL_CERT_DIR'] = path_to_certs`, `Excon.defaults[:ssl_ca_file] = path_to_file`, `ENV['SSL_CERT_FILE'] = path_to_file`, `Excon.defaults[:ssl_verify_callback] = callback` (see OpenSSL::SSL::SSLContext#verify_callback), or `Excon.defaults[:ssl_verify_peer] = false` (less secure).`
```

Any idea ? do i need to add some certificates in the openstack tag in bosh.yml?

**Comment by voelzmo** [comment]
@MoizArif Connectivity issues was what I suspected. For configuring self-signed certificates: the documentation has some information on how to add support for validating self-signed endpoint certs [1]. 

Hope that helps!

[1] http://bosh.io/docs/openstack-self-signed-endpoints.html

**Comment by MoizArif** [comment]
@voelzmo 
After changing the bosh.yml file to include ca_cert, now the `bosh-init deploy bosh.yml` command is failing with:

```
Command 'deploy' failed:
  creating stemcell (bosh-openstack-kvm-ubuntu-trusty-go_agent 3262.2):
    CPI 'create_stemcell' method responded with error: CmdError{"type":"Bosh::Clouds::CloudError","message":"Unable to connect to the OpenStack Keystone API https://public.fuel.local:5000/v2.0/tokens\nSSL_connect returned=1 errno=0 state=SSLv3 read server certificate B: certificate verify failed (OpenSSL::SSL::SSLError) Unable to verify certificate. This may be an issue with the remote host or with Excon.Excon has certificates bundled, but these can be customized.`Excon.defaults[:ssl_ca_path] = path_to_certs`, `ENV['SSL_CERT_DIR'] = path_to_certs`, `Excon.defaults[:ssl_ca_file] = path_to_file`, `ENV['SSL_CERT_FILE'] = path_to_file`, `Excon.defaults[:ssl_verify_callback] = callback` (see OpenSSL::SSL::SSLContext#verify_callback), or `Excon.defaults[:ssl_verify_peer] = false` (less secure).","ok_to_retry":false}
```

to be clear, this is the i am adding the contents of `/etc/keystone/ssl/certs/ca.pem` under the ca_cert tag in bosh.yml. Is it fine?

**Comment by MoizArif** [comment]
@voelzmo 
For now, i have used `ssl_verify_peer: false` in the openstack.connection_options in bosh.yml and now the stemcell is uploaded successfully. But the progress still shows 96%. What do you think can be causing this now?

![image](https://cloud.githubusercontent.com/assets/3896078/16801625/4c92ca10-4915-11e6-956e-98954c49aa55.png)

**Comment by MoizArif** [comment]
Same issue while `bosh upload release` @ 96%

![image](https://cloud.githubusercontent.com/assets/3896078/16802017/a78cf2ae-4917-11e6-8d74-16a131860740.png)

**Comment by voelzmo** [comment]
@MoizArif The progress indicator sticking at 96% although the upload was successful is a known UI bug. As long as the task finished successfully, you're good to go.

Closing this one for now. Please re-open or create a new ticket if you encounter additional problems.

**Comment by kenfting** [comment]
@MoizArif, Hi, Moiz, I am using OpenStack Kilo to deploy cf, and facing the same quesion you have resolved. As you said, I have used ssl_verify_peer: false, connect_timeout: 360. But at last, it's still:

Started update stemcell > Uploading stemcell bosh-openstack-kvm-ubuntu-trusty-go_agent/3262.14 to the cloud. Failed: Unable to connect to the OpenStack Keystone API http://192.168.0.3:5000/v2.0/tokens
Connection timed out - connect(2) for 192.168.0.3:5000 (Errno::ETIMEDOUT) (00:10:42)

I found that you used the keystone public url as auth_url under openstack tag in bosh.yml. Could you tell me that how to get and set public url on openstack? Almost the public url is http://ctl:5000/v2.0 and /etc/hosts could tell us the IP of ctl. Thanks very much!

**Comment by MoizArif** [comment]
Hi @kenfting , I used the same approach you mentioned. Just reiterating, check your /etc/keystone/keystone.conf file for "public_endpoint" and use it as auth_url in the bosh.yml file. If the value is like "https://xxxx:5000/v2.0" use it but make sure that your machine can resolve xxxx into an IP address that was in my case on the Public network. Hope this helps.

**Comment by kenfting** [comment]
Hi, @MoizArif , Thanks very much for your reply. But I found my /etc/keystone/keystone.conf file for "public_endpoint" is <None>, should I set the value to like http:ctl:5000/v2.0 to it? I tried to do this, but it didn't work. I Installed the  BOSH on a VM instance on OpenStack, this IP of ctl is pingable, So I am thinking maybe it is security settings or something, what about username and api_key ? Thanks again!

**Comment by MoizArif** [comment]
Your openstack section of bosh.yml should look like this:
```
    openstack: &openstack
      auth_url: https://public.fuel.local:5000/v2.0
      tenant: <opensack project/tenant name>
      username: <openstack username>
      api_key: <openstack password>
      region: <openstack region name>
      default_key_name: <keypair name>
      default_security_groups: <security group for bosh>
      connection_options:
        connect_timeout: 3600
        ssl_verify_peer: false
```

Also make sure your security group for bosh has the required rules added. [https://bosh.io/docs/init-openstack.html#security-groups](url)

**Comment by kenfting** [comment]
Hi @MoizArif 
Thanks for your reply.
I used keystone v3. the username and api_key from /root/setup/admin-openrc.sh, they are like:
export OS_USERNAME=adminapi
export OS_PASSWORD=e877b22f01ec********
I also tried to use the openstack horizon's username (admin) and password (0e5e********), All these two can create the Director VM and be pingable.  I used ssh vcap＠ip -i ./bosh.pem to check that the private key is correct. In VM, I can not use ping, because I don't know the root's password. But I used curl http:// 130.127.***.***:5000/V3. It prints:
{"version": {"status": "stable", "updated": "2015-03-30T00:00:00Z", "media-types": [{"base": "application/json", "type": "application/vnd.openstack.identity-v3+json"}], "id": "v3.4", "links": [{"href": "http://130.127.***.***:5000/v3/", "rel": "self"}]}}
As you said, I also checked the required rules to be added to security group "bosh" and "cf". 
All the above steps I tried to do, but no luck. it still:
Started update stemcell > Uploading stemcell bosh-openstack-kvm-ubuntu-trusty-go_agent/3262.14 to the cloud. Failed: Unable to connect to the OpenStack Keystone API http://130.127.1**.1**:5000/v3/auth/tokens
getaddrinfo: Name or service not known (SocketError) (00:00:09)

Error 100: Unable to connect to the OpenStack Keystone API http://130.127.1**.1**:5000/v3/auth/tokens
getaddrinfo: Name or service not known (SocketError)

Thanks all the same. I will keep trying another ways

**Comment by MoizArif** [comment]
Can you check the auth_url with using https and see if it works. Assuming the endpoints are configured using https.

**Comment by kenfting** [comment]
After using https, and bosh-init deploy bosh.yml command:
Command 'deploy' failed:
  creating stemcell (bosh-openstack-kvm-ubuntu-trusty-go_agent 3262.12):
    CPI 'create_stemcell' method responded with error: CmdError{"type":"Bosh::Clouds::CloudError","message":"Unable to connect to the OpenStack Keystone API https://130.127.1**.1**:5000/v3/auth/tokens\nSSL_connect returned=1 errno=0 state=SSLv2/v3 read server hello A: unknown protocol (OpenSSL::SSL::SSLError)","ok_to_retry":false}

**Comment by MoizArif** [comment]
are you sure you are using keystone api version v3 ? check by using "keystone endpoint-list | grep 5000" command on your OpenStack controller to check your public & internal endpoint urls for keystone.

**Comment by kenfting** [comment]
Yes, it is keystone v3.
# openstack endpoint list | grep 5000
| 0eeca224bc044ac1bb816b2cb43a8034 | RegionOne | keystone     | identity        | True    | public    | http://ctl:5000/v3                    |
| 4152723e7c654864925547213b80dfd2 | RegionOne | keystone     | identity        | True    | internal  | http://ctl:5000/v3 

So the public and internal are the same
